proposals/permissions-requestSiteAccess-api.md

```diff
@@ -0,0 +1,135 @@
+# permissions.requestSiteAccess() API
+
+## API Overview
+
+### Background
+
+Extensions can request host permissions via:
+- Manifest declared `host_permissions`, granted at install time by default
+- Manifest declared `optional_host_permissions`, granted at runtime after a user gesture by `<browser>.permissions.request()`
+- Manifest declared `matches` for a `content_script`, granted at install time by default
+
+Users can withhold host permissions, causing the extension to only have access when the extension is [invoked](https://developer.chrome.com/docs/extensions/develop/concepts/activeTab#invoking-activeTabhttps://developer.chrome.com/docs/extensions/develop/concepts/activeTab#invoking-activeTab).
+
+Each browser may decide how it signals the user when an extension is requesting access (e.g through the extensions menu). However, there is no way for an extension to explicitly signal at runtime it’s requesting site access after it was withheld without a user gesture and a heavy-weight permission dialog.
+
+
+### Objective
+
+Allow the extension to show site access requests at runtime without any user gesture in a less-obtrusive way than with permissions.request(). This can be done with a new API that:
+
+- Applies to a specific tab or document id
+- Doesn’t need to be made inside the handler for a user action
+- Shows the request in the UI, handled differently by each browser. See more in [UI Elements and User-Visible Effects](#ui-elements-and-user-visible\-effects) section
+- When accepted, grants always access to the site’s origin
+- Resets the request on cross-origin navigation
+
+#### Use Cases
```

Comment:
Overall, this is a confusing API to developers, and it would be better if there were more concrete use cases. What value can this API bring to extensions? Is there any relevant data for browsers' "site access" function? Are users satisfied with "site access" function, or are users encountering any problems? The above questions are just to avoid developing an API that no one uses.

Reply:
I can share a bit of extra context here. We are still looking at possible permission UI changes like the ones mentioned at [I/O](https://youtu.be/QYd2XiUYNlE?t=608). This doesn't include changing any defaults, but we do still expect more users to experiment with different access settings. It feels like putting APIs in place first is better than adding them retrospectively, which would imply a gap between when the UI launches and when APIs to adapt to them are available.

Reply:
This feels like it belongs in a manifest rather than as an API. I think extensions fall into these categories:

- on demand: just take an action when invoked (`activeTab`)
- for specific websites: think "Reddit Enhancement Suite" or "Refined GitHub", that have a pre-set list of domains to run on
- for all websites: think content blockers, Grammarly

Between these there are a couple more cases:

- A: pre-set list, but more can be added
- B: all websites, but some can be disabled

"A" can be achieved via `.request()` when the user desires; "B" is covered by Chrome's current UI.

I'm not sure where `requestSiteAccess` falls on this spectrum.

Once I specify `optional_host_permissions: all_urls`, the browser should already:

- Show "This extension is available on this website"
- Allow toggling the permission

---

As a user, I wish that I had a way to enable the type A on more websites in a consistent way.

As a developer, I wish I didn't have to manually handle content script registration and permission requests.

Reply:
I definitely agree that not having to manually handle this would be nice and think that'd be interesting to experiment with. For some of your cases I agree that might work.

I do think there is another case to the ones you mentioned. Consider a shopping extension that can only offer coupons on certain sites, and it downloads this list remotely. It specifies `<all_urls>`, so can run everywhere, but a user explicitly revokes its access from all pages. It would be nice if the extension had a way to less intrusively flag "I checked my list and I can run here, you might want to add me back on this domain".

Reply:
>  It would be nice if the extension had a way to less intrusively flag "I checked my list and I can run here, you might want to add me back on this domain".

Why isn't the list in the manifest? 

But still from a manifest standpoint, what is the point of `host_permissions` if they can be disabled? It sounds like we need:

- compatible host list: all_urls
- suggested host list: eBay, amazon, etc

To fit into this proposal:

- compatible host list: `host_permissions`
- suggested host lists: a new `preferred_hosts`

The drawback is that extension developers have to submit a new extension to update the list, but the advantages are that the browser has total control on how/when to show it, and the developer doesn't have to deal with issues like "when and how often can I `requestSiteAccess`?"

Reply:
As-is the current proposed API would not be useful in Safari for shopping extensions, since the URLs are hidden from extensions without host permissions. A declarative model might be the better approach.

With declarative approach, shopping like extensions to register patterns and reasons / descriptions to say "I'm interested in shopping.example/* because I have '2 coupons for 10%-20% off'". This would give a signal the browser can bubble up in the UI to the user in a privacy preserving way, that has a call to action for why the user would want to grant access.

Reply:
In addition to what others have already mentioned, host permissions are not always bound to Tabs. For example, an extension needs a host permission, just to bypass CORS limitation in the background page. User never opens a tab for that host.

And host permissions are not always 1-to-1 relationship to Tab's URL. It may also be a 1-to-many or many-to-many relationship. For example, an extension needs two host permissions, only granting one by tab's host still can't work.

In short, host permissions are more complex and can't be reduced to the Tab's host the user is opening. If users want to manage host permissions fine-grained manually, existing functionality (in the chrome://extensions/?id=extension-id) already supports this.

Reply:
Users often don't want to manage the host permissions though, and will just grant access to all sites to avoid any hassle. We should strive to make the defaults be privacy preserving and not require host permission micro-management. So new APIs should be considered that improves things for the user and developers.

Reply:
Thanks everyone for your comments. I'll try to summarize this.

There are two main **issues**:
1. URLs are hidden from extensions without host permissions
2. Host permissions are not always bound to tab

There is one main **suggestion**:  declarative approach by listing the sites in the manifest

For 1., this is a browser design constraint and something we should talk about in more details.. which will be easier to do in the next WECG meeting

For 2., if I understand this correctly the multiple host permissions can happen when there are multiple iframes with different host permissions. The request would only be for the top origin, same as how `activeTab` works

For the declarative approach suggestion, the main reason we don't want to do that is that we want there to be a deliberate, strong signal that the extension can do something useful on the site and really wants to run.  This should not be a passive, "hey, I think I can do something here", it should be a "hey, you, the user, probably want me to do something here".  

When we talked internally at Chrome about having this be declarative and take in URL patterns (either via a runtime method or a manifest field), it became far too close to host permissions themselves. We suspect that the vast majority of extensions would just have the same field match as in their host permissions, since there is no more knowledge at manifest time about why the extension would run on a specific site.

For example, for the shopping extension it can request <all_urls> host permission (or even a large set of specific site) but only show the request when there is a coupon (e.g it doesn't need to show on store.com but yes on store.com/item)

Yes, this is more work for developer since they need to call it every time. However, we hope this to be better for the extension, it can send requests when it really wants to run, and the user, more likely to accept the requests.

Reply:
Now answering some specific comments

@fregante 
> In your proposed new UI, it looks like this could be covered already: just show the extension on the list, but disabled.

This is covered in the UI, but wasn't clear. Extensions with withheld host permissions, or extensions with not-yet granted optional host permissions are listed in the extensions menu and display they are requesting access.
The proposal here is giving the extension a way to request site access on a site that doesn't require user action

> Why isn't the list in the manifest?

I believe this was answered in previous comment :)

> The drawback is that extension developers have to submit a new extension to update the list, but the advantages are that the browser has total control on how/when to show it, and the developer doesn't have to deal with issues like "when and how often can I requestSiteAccess?"

With both approaches, the browser would have control on how/when to show it. Similarly, with both approached developer can only request site access if it has host permission for the site. As of how often, it is deliberately left as open ended in case browsers need to mitigate abuse scenarios (which would be properly communicated if implemented)

@xeenon, we can discuss further in WECG about the URL constraint in Safari

@hanguokai, I believe I answered the multiple host permissions on a tab on previous comment

Reply:
These are good point to bring up. Happy to discuss them further at the next public WECG meeting, which is tomorrow February 1st :)

Reply:
As oliver said before, my understanding is that the background for the use of this API is that when the extension is installed, it is not granted any host permissions although it declares (non-optional) host permissions in manifest.json.

In addition, I feel you are implying that this API is only for content scripts, not for extension contexts (service worker, extension pages, etc). And the function only runs when the page is visually to the user. If so, this is obviously only some use cases of extensions, and it is not enough to address other cases where extensions do not have permissions after installation. 

> For 2., if I understand this correctly the multiple host permissions can happen when there are multiple iframes with different host permissions. The request would only be for the top origin, same as how activeTab works

Not exactly that case, for example, 
1. The network requests of many large websites are complex. After the page is loaded, it will request data from other hosts. E.g. when "www.site-A.com" is loaded, then it fetches data from "api.site-A.com". In this case, the extension may need "api.site-A.com" permission instead of "www.site-A.com" permission in the service worker for it's function. The user never open "api.site-A.com" in a tab, because this domain is not designed for users.
2. If some core host permissions (not related to a specific shopping website) are not granted, the extension may can't work (i.e. `host_permissions` in manifest.json).

Reply:
> In addition, I feel you are implying that this API is only for content scripts, not for extension contexts (service worker, extension pages, etc). And the function only runs when the page is visually to the user.

This is definitely the intent. I understand it doesn't solve background use cases - since this is a very high visibility prompt we want it to only show when linked to an open tab that the user has context for. Improving the background use case is a non-goal but that isn't to say we couldn't look into new APIs in the future if needed.

Reply:
I can't reconcile these two statements:

> It specifies `<all_urls>`, so can run everywhere, but a user explicitly revokes its access from all pages

> this API is only for content scripts

Content scripts cannot run if the user revoked a permission. It seems that this API exclusively fits this situation:

1. the extension has permission to execute scripts in a tab (temporarily or permanently)
2. it asks for further host permissions or it asks for permanent permission to the current context

Seems niche, why not just have the extension deal with this? This way they can clearly communicate the intent and reason behind the request.

As a user, if a the browser asks for a permission for a third domain, I don’t think I'd accept without further explanation, especially if I explicitly blocked the extension on that domain before.

Reply:
Ah, the wording may have been slightly confusing there. It would often be used to _grant access_ to content scripts. The API itself would definitely be available in background contexts.

An example use case in Chrome would be something like this:

- An extension requests <all_urls> in host_permissions
- The user changes the site access to "on click" or "on specific sites" (note that they can do this even if the host permissions weren't specified as optional)
- In the background, the extension realises it can inject a useful content script (for example, it notices a tabs.onUpdated event and the tab has a URL it has special logic for)
- The extension would then prompt for access to that tab ID

Does that make sense?

Reply:
Let's take a step back and see current behavior (updated proposal `Background` section too)
_Please, feel free to correct me on other browsers behavior since I'm not fully familiar!_

- An extension can request host permissions. At installation, they are granted (Chrome, Firefox) or withheld (Safari) by default
- An extension can request optional host permissions, that can be granted after a user action
- In Chrome and Safari, extension site access can be withheld
- In Chrome, withholding an extension site access means the extension is set to run "on click". When the extension is clicked (either the pinned action or through the extensions menu) it gets *access to run on the origin top frame until cross-origin navigation*.
- `permissions.request()` allows an extension to request site access after a user action

Let's have an example now (similar to Oliver's on previous comment). In Chrome:
1. An extension requests `<all_urls>` in host_permissions
2. User navigates to site.com
3. The user changes the `site.com` site access to "on click"
4. User navigates to `site.com/useful`
5. In the background, the extension realizes it can inject a useful content script on `site.com/useful` 

User can realize the extension wants to run by opening the extensions menu. However, there is no way for the extension to say "I really want to run on this specific page" without user action. Yes, it could work with a declarative list. However, we have listed our reason why we prefer not to (updated proposal's `Other Alternatives Considered` section)

@hanguokai 
As for the issue of extension requesting host_permissions that are not in the tab top frame origin. At least on Chrome, this is the actual behavior of granting site access to an extension. Granting site access to a host (either by activeTab or allowing site access "on site" / "on all sites") grants only to that host, meaning the tab's top frame origin. This new API is following the same, so it wouldn't fix that issue.

@fregante 
> it asks for further host permissions or it asks for permanent permission to the current context

Extension cannot ask further host permissions. It would ask for host_permissions that were withheld by the user

> As a user, if a the browser asks for a permission for a third domain, I don’t think I'd accept without further explanation, especially if I explicitly blocked the extension on that domain before.

Browser cannot ask permission for a third domain

Reply:
Thanks for your further explanation. Now I have a clearer understanding of the intent of this API. But in my opinion, this is still a controversial API.

## How to use this API

> 5. In the background, the extension realizes it can inject a useful content script on site.com/useful

Can you further explain what developers are supposed to do here? 

Because this API is bound to browser tabs, I think developers need to use `tabs.onCreated/onUpdated/onReplaced` events to detect tab's top frame changes, and must need the "tabs" permission to get `tab.url` and/or `tab.title`. In these events, developers are facing two possible situations: 1. already has the host permission and already inject content scripts 2. don't has the host permission and no injected content scripts. Therefore, developers first need to determine whether has the host permission. If no host permission, at this stage, extensions only know tab's url and title, and **don't know the page's content** since it has not inject a content script. At this point, developers can only rely on these two pieces of information (url and title) to determine whether to apply for the site permission, i.e. call `permissions.requestSiteAccess()`.

Here, without knowing the content of the page, it may be difficult for developers to determine whether it strongly requires the site access. In the case of shopping, I only know the URL and title, I may can't know what product is being sold on the current page, so I can't tell if there is a coupon for that product.

## The model itself

> In Chrome, withholding an extension site access means the extension is set to run "on click". When the extension is clicked (either the pinned action or through the extensions menu) it gets access to run on the origin top frame until cross-origin navigation.

I rarely use this feature. I also rarely hear users use this feature. So I want to repeat the question I asked at the very beginning: Is there any relevant data about this function? Are users satisfied with this function, or are users encountering any problems at present?

I think the model itself is flawed, it only fits a small subset of real use cases, and a lot of use cases (perhaps 50% or more) don't fit the model (We have already given examples before).

## Experiment with this API without standardization

In my opinion, there are still doubts about the actual value of this API for developers and users. I think Chrome can first experiment with this API (for a year) to see the actual effect. Tell developers explicitly that this is an experimental API and that it may be modified incompatibly or even removed in the future. This way, an experimental API does not require the agreement from other browsers. For example, let developers apply an Origin Trial token to experiment this API (see #454).

Reply:
### How to use this API

Correct, those are the two scenarios although I would phrase them slightly differently (to match terminology):
a. Extension request host permission and it's granted by the user. Extension has injected scripts
b. Extension request host permission and it's withheld by the user. Extension ha no injected  scripts

On scenario a., you are right. Extension does not know the page content.  Extension has access to:
- `permissions.getAll()` returns the active permissions. Extension can know if a host permission is not in the returned permissions origin
- `tabs.get()` returns tab URL, title, ...
- other APIs like `windows.getCurrent()` to get more information 

> Here, without knowing the content of the page, it may be difficult for developers to determine whether it strongly requires the site access. In the case of shopping, I only know the URL and title, I may can't know what product is being sold on the current page, so I can't tell if there is a coupon for that product.

Currently, the only way for developers to determine whether it requires site access is to check if it has active permission for the page. This provides a way to show a visible request when this happens. 

Yes, it's probably not sufficient to know given a tab id / document id that the extension "strongly needs access to the site". However, what would be another way to determine that? (if host permissions are withheld, extension doesn't have access to the page). By looking at document id / tab id, extension can get more information than if it just gave a declarative list of URLs (as explained in previous point).  

### The model itself

> I rarely use this feature. I also rarely hear users use this feature. 

User withholding host permissions is not a common practice today. However, we expect that to change in the future because:

- Browsers are exploring withholding permissions at installation. Safari already does that, and I believe Firefox is exploring that in MV3. Chrome is exploring that too
- At least in Chrome, it will be more accessible for the user to withhold host permission in the extensions menu (that means, setting the extension to run "on click")

> Is there any relevant data about this function? Are users satisfied with this function, or are users encountering any problems at present?

We did user research on showing a request in the Chrome toolbar every time a extension wants site access to a withhold host permission. Users found it to be too nosy, and not always helpful. However, they mentioned that the request is relevant then it provides value.
Thus, we drafted this proposal to show the request only when the extension requests to show it. Yes, it can be always but at least it gives the option to the extension to have it be more relevant and the browser to better control it. 

> I think the model itself is flawed, it only fits a small subset of real use cases, and a lot of use cases (perhaps 50% or more) don't fit the model (We have already given examples before).

Agree, the real use cases subset is small now. However, we expect this to increase as mentioned on the first point on this section.

### Experiment with this API without standardization

Do other browsers see a potential path forward for this API? Otherwise, Chrome can continue without standardization

Reply:
Proposal was updated to include url so it can fit other browsers. Browser can add requirements on which parameters must be provided. For example, Chrome will require `tabId` or `documentId` to be provided and `url` can be added on top of that.